packages/devtools_app/lib/src/shared/primitives/utils.dart

```diff
@@ -442,84 +442,101 @@ class RateLimiter {
   }
 }
 
+// If the need arises, this enum can be expanded to include any of the
+// remaining time units supported by [Duration] - (seconds, minutes, etc.).
 /// Time unit for displaying time ranges.
-///
-/// If the need arises, this enum can be expanded to include any of the
-/// remaining time units supported by [Duration] - (seconds, minutes, etc.). If
-/// you add a unit of time to this enum, modify the toString() method in
-/// [TimeRange] to handle the new case.
 enum TimeUnit { microseconds, milliseconds }
 
-class TimeRange {
-  TimeRange({this.singleAssignment = true});
+/// A builder used to build a well-formed [TimeRange] incrementally.
+final class TimeRangeBuilder {
+  /// Creates a new [TimeRangeBuilder] to build a [TimeRange]
+  /// with the optionally specified [start] and [end] initial values.
+  TimeRangeBuilder({int? start, int? end}) : _start = start, _end = end;
 
-  factory TimeRange.offset({
-    required TimeRange original,
-    required Duration offset,
-  }) {
-    final originalStart = original.start;
-    final originalEnd = original.end;
-    return TimeRange()
-      ..start = originalStart != null ? originalStart + offset : null
-      ..end = originalEnd != null ? originalEnd + offset : null;
+  int? _start;
+
+  /// Sets the start time of this builder.
+  ///
+  /// The start time should be less than or equal to the end time.
+  set start(int startTime) {
+    _start = startTime;
   }
 
-  final bool singleAssignment;
+  int? _end;
 
-  Duration? get start => _start;
+  /// Sets the end time of this builder.
+  ///
+  /// The end time should be greater than or equal to the start time.
+  set end(int endTime) {
+    _end = endTime;
+  }
 
-  Duration? _start;
+  /// Whether both `start` and `end` properties are set,
+  /// meaning [build] can safely be called.
+  bool get canBuild => _start != null && _end != null;
 
-  set start(Duration? value) {
-    if (singleAssignment) {
-      assert(_start == null);
+  /// Returns a [TimeRange] built from the specified [start] and [end] values.
+  ///
+  /// If either [start] or [end] is `null`, throws an error.
+  ///
+  /// The [start] time must be less than or equal to the [end] time.
+  TimeRange build() {
+    final startDuration = _start;
+    if (startDuration == null) {
+      throw StateError('TimeRangeBuilder.start must be set before building!');
```

Comment:
I think an assertion is better here than a `StateError`, since this would indicate an internal bug with how we are calling this within the code

Reply:
I believe that's what [`StateError`](https://api.dart.dev/dart-core/StateError-class.html) is for, indicating this object is in a state not appropriate for calling `build`. For example `list.first` also throws a `StateError` if there's nothing in the list. 

Besides that an assertion won't promote the value to be non-null, so we need to add not-null assertions as well.

I kept it as is for now, but let me know if you'd like me to make further adjustments.